GitUI/UserControls/RevisionGrid/Graph/ThreadSafeGrowingListEnumerator.cs

```diff
@@ -0,0 +1,38 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+
+namespace GitUI.UserControls.RevisionGrid.Graph
+{
+    public class ThreadSafeGrowingListEnumerator<T> : IEnumerator<T>
```

Comment:
It's possible to have a non-allocating `foreach` if this enumerator type is a `struct` and the `GetEnumerator` method returns the struct type. You'd have to make the `IEnumerable<>` version explicit too to avoid boxing.

Look at `List<T>` for reference.

Reply:
I took the advise of @sharwell and changed this to:
`return _internalArray.Take(Count).GetEnumerator();`
The custom enumerator class is now deleted.

Reply:
How frequently are you enumerating this collection?

Reply:
Very frequent... Multiple times for every commit (the collections usually only have 1 or 2 items). But only during the loading phase. Memory and performance are important since we want repositories to load as quick as possible..:) Suggestions are welcome!

Reply:
In that case you should consider a struct based enumerator. See how [this type](https://github.com/dotnet/project-system/blob/master/src/Microsoft.VisualStudio.ProjectSystem.Managed/Text/LazyStringSplit.cs) implements its enumerator. It's possible to `foreach` over a that type without allocating anything for the enumerator, which is pretty nice. In fact your type doesn't even need to implement `IEnumerable<T>`. Just have a `GetEnumerator` method that returns your struct.

The struct will have a readonly reference to the array, the readonly count and a mutable index field to update in `MoveNext`.

Reply:
For loops might be faster still. You'd have to benchmark, but I think RyuJIT only optimises away the bounds check if you use the `for (int i = 0; i < array.Length; i++)` pattern (i.e. using `.Length` in the condition). Even with the bounds check, it's less machinery than going through enumerable.

If it's a hard change, you might like to run a microbenchmark with BenchmarkDotNet on a smaller and similar code fragment to see what kind of difference you're likely to see.

Reply:
I replaced the enumeration with a for loop, where it really matters. Also replaced the Any(), so no enumeration is needed. This is just a minor improvement. Overall the improvement is noticable. In the Linux repo, when the full history is loaded, the difference is huge.